### DISCUSSION TITLE
No effect from WebP's `near-lossless` option in ImageMagick convert

### DISCUSSION THREAD

**Discussion by alanorth** [ImageMagick/ImageMagick#6204] [Help] [answered]
### ImageMagick version

7.1.1.5-1

### Operating system, version and so on

Arch Linux

### Description

The [ImageMagick docs for WebP](https://imagemagick.org/script/webp.php) note that there should be a "near-lossless" option that maps to cwebp's `-near_lossless`. I tried to test that briefly and it doesn't seem to be working:

```console
$ for q in 20 40 60 80 90; do magick convert data/10568-103447.pdf\[0\] -define webp:near-lossless=$q /tmp/10568-103447-near-lossless-q${q}.webp; done
$ ls -l /tmp/10568-103447-near-lossless-q*
-rw-r--r-- 1 aorth aorth 37470 Mar 29 14:23 /tmp/10568-103447-near-lossless-q20.webp
-rw-r--r-- 1 aorth aorth 37470 Mar 29 14:23 /tmp/10568-103447-near-lossless-q40.webp
-rw-r--r-- 1 aorth aorth 37470 Mar 29 14:23 /tmp/10568-103447-near-lossless-q60.webp
-rw-r--r-- 1 aorth aorth 37470 Mar 29 14:23 /tmp/10568-103447-near-lossless-q80.webp
-rw-r--r-- 1 aorth aorth 37470 Mar 29 14:23 /tmp/10568-103447-near-lossless-q90.webp
```

The file sizes are all the same.

I tried again with cwepb (from libwebp 1.3.0) and it works as expected:

```console
$ magick convert data/10568-103447.pdf\[0\] -quality 100 img/10568-103447.pdf.jpg
$ for q in 20 40 60 80 90; do cwebp -near_lossless $q img/10568-103447.pdf.jpg -o /tmp/10568-103447-cwebp-near-lossless-q${q}.webp; done
$ ls -l /tmp/10568-103447-cwebp-near-lossless-q*
-rw-r--r-- 1 aorth aorth 115394 Mar 29 14:26 /tmp/10568-103447-cwebp-near-lossless-q20.webp
-rw-r--r-- 1 aorth aorth 122056 Mar 29 14:26 /tmp/10568-103447-cwebp-near-lossless-q40.webp
-rw-r--r-- 1 aorth aorth 118732 Mar 29 14:26 /tmp/10568-103447-cwebp-near-lossless-q60.webp
-rw-r--r-- 1 aorth aorth 140514 Mar 29 14:26 /tmp/10568-103447-cwebp-near-lossless-q80.webp
-rw-r--r-- 1 aorth aorth 140514 Mar 29 14:26 /tmp/10568-103447-cwebp-near-lossless-q90.webp
```

I see a [related question about this](https://github.com/ImageMagick/ImageMagick/discussions/3029) from a few years ago, but I'm not sure which ImageMagick version shipped with support for this. Thanks!

**Comment by fmw42** [comment]
What version of webp is installed with Imagemagick 

```
convert -list format | grep webp

     WEBP* WEBP      rw+   WebP Image Format (libwebp 1.3.0 [020F])
```

**Reply by alanorth** [reply]
Hey @fmw42, here's what is reported for me:

```console
$ convert -list format | grep webp
     WEBP* WEBP      rw+   WebP Image Format (libwebp 1.3.0 [020F])
```

**Reply by fmw42** [reply]
Looks OK to me. But I am not an expert on WEBP. So wait for one of the IM developers to comment.

**Comment by urban-warrior** [comment, marked as answer]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/archive/beta/ by sometime tomorrow. 

Use `-define webp:lossless=1 -quality 90`, for example.

**Reply by alanorth** [reply]
Great, thank you! Regarding your comment and my naive reading of [the patch](https://github.com/ImageMagick/ImageMagick/commit/3331c9ed8bdbbdc548f8309a3fadf5b6e638bc98):

> _Use -define webp:lossless=1 -quality 90, for example._

What is the relationship between `webp:lossless` and the webp encoder's `near_lossless` now? I see the [ImageMagick WebP docs](https://imagemagick.org/script/webp.php) have removed the mention of `webp:near-lossless` since yesterday.

Furthermore, from reading [`coders/webp.c` on main as of this patch](https://github.com/ImageMagick/ImageMagick/blob/3331c9ed8bdbbdc548f8309a3fadf5b6e638bc98/coders/webp.c#L1143) I don't understand the interaction between `webp:lossless`, `-quality`, and the webp encoder's `near_lossless`.

**Comment by alanorth** [comment]
Thank you for the patch, @urban-warrior. From reading [the `coders/webp.c` code](https://github.com/ImageMagick/ImageMagick/blob/3331c9ed8bdbbdc548f8309a3fadf5b6e638bc98/coders/webp.c#L1143), ImageMagick seems to _always_ configure `near_lossless` mode using the given `-quality` if the cwebp encoder ABI supports it:

```c
  if (image->quality != UndefinedCompressionQuality)
    {
      configure.quality=(float) image->quality;
#if WEBP_ENCODER_ABI_VERSION >= 0x020e
      configure.near_lossless=(float) image->quality;
#endif
    }
```

Thanks

**Comment by urban-warrior** [comment]
The theory is that the near lossless parameter is a no-op unless lossless compression is enabled.  Our testing confirms our conjecture.
